### PR TITLE
Semaphores for ocaml probes

### PR DESCRIPTION
**Pull request by gretay-js**
Add separate space for a separate semaphore for ocaml-probes.  Ocaml probes should not use systemtap semaphores because they control different handlers.

The PR also tidies semaphores symbols that need not be global.

### CODE REVIEW

**Comment by gretay-js** [comment]
I've rebased this PR and tested it using a separate tracer library. 
To ensure that the tracer works correctly for executables built with both original and new versions of probes, we can record the version "provider" field of stapsdt notes. The "provider" value was "ocaml" in the original version. I added a  commit to this PR that set "provider" to "ocaml.1" for the new version. I think this internal name is unlikely to be confused with the "official" ocaml version.

File: backend/amd64/emit.mlp [hunk 1]
```diff
@@ -944,7 +944,11 @@ let emit_instr fallthrough i =
        semaphores are of type unsigned short.
        [1] https://sourceware.org/systemtap/wiki/UserSpaceProbeImplementation
     *)
-    I.mov (addressing (Ibased(semaphore_sym, 0)) WORD i 0) (res16 i 0);
+    (* OCaml has it's own semaphore, in addition to system tap,
+       to control ocaml probe handlers independently from stap probe handlers.
+       It is placed immediately after stap semaphore, and is the same
+       size - hence offset 2. *)
```
Comment (mshinwell) [inline backend/amd64/emit.mlp:950]: I don't understand why this is offset 2 not 1.  Is this value not in units of 8-byte words?

**Review by mshinwell — CHANGES_REQUESTED** [review]

File: backend/amd64/emit.mlp (hunk 1 above)
Comment (gretay-js) [reply backend/amd64/emit.mlp:950]: The units are bytes, and the size of the new semaphore is set to be the same as the size of stap semaphores, so the code for reading them remains unchanged.

File: backend/amd64/emit.mlp [hunk 2]
```diff
@@ -1421,7 +1425,7 @@ let emit_probe_notes0 () =
       D.qword (const 0)
     end;
     D.qword (ConstLabel semaphore_label);
-    D.bytes "ocaml\000";
+    D.bytes "ocaml.1\000";
```
[the remark this answers is not in the document]
Comment (gretay-js) [reply backend/amd64/emit.mlp:1428]: ah, suggestions for a better name are most welcome!
The prefix ".1" is the version number for ocaml probe notes. There is already a proposal for "ocaml.2"  in PR #142.
The suffix "ocaml" is useful for distinguishing probes emitted by  OCaml compiler from other stap "providers" of probes.

[the remark this answers is not in the document]
Comment (gretay-js) [reply backend/amd64/emit.mlp:1428]: Users will see these strings if they inspect probe notes using "readelf -n" or gdb's "info probes".

File: backend/amd64/emit.mlp (hunk 1 above)
Comment (mshinwell) [reply backend/amd64/emit.mlp:950]: ah, and the semaphores are 16-bit quantities.  ok.

File: backend/amd64/emit.mlp (hunk 2 above)
[the remark this answers is not in the document]
Comment (mshinwell) [reply backend/amd64/emit.mlp:1428]: This seems ok for now.

**Review by mshinwell — APPROVED** [review]

**Pull request merged**